// nativescript-preview-sdk/src/services/helpers-service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var shortId = require("shortid");
var nodeBtoa = require("btoa");
var HelpersService = /** @class */ (function () {
    function HelpersService() {
    }
    HelpersService.prototype.getHiddenProperty = function () {
        var prefixes = ['webkit', 'moz', 'ms', 'o'];
        if ('hidden' in document) {
            return 'hidden';
        }
        for (var i = 0; i < prefixes.length; i++) {
            if ((prefixes[i] + 'Hidden') in document)
                return prefixes[i] + 'Hidden';
        }
        return null;
    };
    HelpersService.prototype.isBrowserTabActive = function () {
        if (!this.isBrowser()) {
            // not in browser
            return true;
        }
        var property = this.getHiddenProperty();
        if (!property)
            return true;
        return !document[property];
    };
    HelpersService.prototype.areCaseInsensitiveEqual = function (value1, value2) {
        return (value1 || "").toLowerCase() == (value2 || "").toLowerCase();
    };
    HelpersService.prototype.shortId = function () {
        return shortId.generate();
    };
    HelpersService.prototype.isBrowser = function () {
        return (typeof window !== "undefined");
    };
    HelpersService.prototype.base64Encode = function (input) {
        if (this.isBrowser()) {
            return window.btoa(unescape(encodeURIComponent(input)));
        }
        else {
            return nodeBtoa(unescape(encodeURIComponent(input)));
        }
    };
    return HelpersService;
}());
exports.HelpersService = HelpersService;
